REF: simplify latex formatting

ivanovmg opened the pull request:
- [x] closes https://github.com/pandas-dev/pandas/issues/35790
- [x] tests added ``pandas/tests/io/formats/test_latex.py``
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Refactor ``to_latex`` using polymorphism and the builder design pattern.

Polymorphism
--------------

Previously there was a complicated logic in multiple methods
for either longtable or regular table.
This PR implements various builders:
  - ``RegularTableBuilder``
  - ``LongTableBuilder``
  - ``TabularBuilder``

Selection of the appropriate builder is carried out in ``LatexFormatter``,
depending on the kwargs provided.

Each builder must implement construction of all the table's components:
  - Beginning and end of the environment,
  - Separators (top, middle, bottom),
  - Header,
  - Body.

This allows one eliminate complex logic throughout multiple methods.

Separate Concerns
------------------

Separated concerns: row formatting into string (separate class ``RowStringConverter``)
is carried out independently of the table builder.
Derived from ``RowStringConverter`` are two classes used to iterate over headers or body of the tables:
  - ``RowHeaderIterator``
  - ``RowBodyIterator``

Extract Methods
----------------

Multiple methods were extracted to improve code readability.

ivanovmg committed:
Refactor to_latex using polymorphism

Previously there was a complicated logic in multiple methods
for either longtable or regular table.

This commit implements
 - ``LatexTableFormatter``,
 - ``LatexTabularFormatter``,
 - ``LatexLongTableFormatter``,
derived from ``LatexFormatter``, based on ``LatexFormatterAbstract``.

Each of the derived classes implements its own methods for writing
 - beginning of the table;
 - caption and labels;
 - separators;
 - end of the table.

LatexFormatter changes
----------------------

 - Make the process of creating tables more readable.
 - Drop escape and bold_rows attr

ivanovmg committed:
Move strcols from property to attr for performance

ivanovmg committed:
Separate row format from LatexFormatting

New classes for iterations over headers and body:
 - RowHeaderIterator;
 - RowBodyIterator,
based on RowCreator.

ivanovmg committed:
Use property for table components

ivanovmg committed:
Enable builder pattern for various types of tables

This reverts ``pandas/io/formats/format.py``
to its nearly original state and enables polymorphism
under the hood inside ``pandas/io/formats/latex.py``.

Add ``pandas/tests/io/formats/test_latex.py`` to test
lower-level functions/classes declared in
``pandas/io/formats/latex.py``

ivanovmg committed:
Fix linting

ivanovmg committed:
REF: simplify logic in _caption_and_label

arw2019 reviewed (changes requested):
thanks for the PR!

According to the code coverage bot under these changes patches of code are not covered by the tests. The thing to do there is for each patch either clean it up or add tests so that they're hit 

Also, if I've got this right you're adding/removing methods. For new methods we'll need docstrings; old methods need to be deprecated prior to removal

ivanovmg commented:
What is not covered by the tests are the abstract methods.

```
    @property
    @abstractmethod
    def env_end(self):
        pass
```
I think that the best thing is to setup coverage in such a way to ignore simple pass in the function.
Alternatively I can simply put a docstring there, which will be considered by the bot as a check.

The methods removed (actually one, ``_write_tabular_begin``) and renamed (``_print_cline`` -> ``_compose_cline``) are internal implementation. Do I need to keep them and mark deprecated? They are not used anywhere else.

ivanovmg committed:
Add docstrings

arw2019 commented on pandas/io/formats/latex.py:
```diff
@@ -10,56 +11,87 @@
-        caption: Optional[str] = None,
-        label: Optional[str] = None,
-        position: Optional[str] = None,
+        self, formatter, multicolumn=False, multicolumn_format=None, multirow=False,
```
looking across to `LatexFormatter` I think we want type hints here

ivanovmg replied:
Added type hints.

arw2019 commented on pandas/tests/io/formats/test_latex.py:
```diff
@@ -0,0 +1,95 @@
+import pytest
```
I think this should go in `pandas/tests/io/formats/test_to_latex.py`

ivanovmg replied:
Moved those tests into ``test_to_latex.py``

ivanovmg committed:
Add type hints to RowStringConverter

ivanovmg committed:
Move tests from test_latex to test_to_latex

arw2019 reviewed (approved):
Re: deprecation I'm not sure. I would think it's not necessary as long as the methods are not exposed to users.

I would maybe change the title of the PR to something like "REF: simplify latex formatting" to make it clearer to other reviews what your goal is.

Otherwise looks okay to me. You'll need feedback from the core team to move forward

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -107,89 +137,59 @@ def pad_empties(x):
+        return crow
 
-        buf.write("\\toprule\n")
+    @staticmethod
```
instead of using staticmethods can you make these module level functions. I find this much harder to follow and non-idiomatic.

ivanovmg replied:
Moved static methods to module level. My intention originally was to keep the methods in the class, since these functions are not to be used elsewhere, in general, and to make it clear where the functions belong to. Anyway, I did as you suggested.

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -10,56 +11,91 @@
+        return self._nlevels if self.fmt.header else 0
+
+    @property
+    def _ilevels(self):
```
i would make more descriptive properties here; they don't really need to be private, but ok

e.g.

index_nlevels is more descriptive.

ivanovmg replied:
Actually, it was not clear for me what these ``_ilevels``-like values were in the original code. So, I decided not to change the names. As you suggested, I renamed them.

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -107,89 +137,59 @@ def pad_empties(x):
 
-        buf.write("\\toprule\n")
+    @staticmethod
+    def _escape_symbols(row):
```
ideally type *every* parameter both in and out

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -199,7 +199,7 @@ def _format_multicolumn(self, row: List[str], ilevels: int) -> List[str]:
         \multicolumn{3}{l}{a} & b & \multicolumn{2}{l}{c}
         """
-        row2 = list(row[:ilevels])
+        row2 = list(row[: self._ilevels])
```
why is this not a list already?

ivanovmg replied:
Indeedl. But I just re-used the original code as it was, with minimum intrusion.

ivanovmg replied:
Fixed that.

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -256,88 +254,394 @@ def _format_multirow(
+    def bottom_separator(self):
+        """Bottom level separator."""
+
+    @property
```
ok with these being properties, but generaly we like functions if ever an intention of adding args.

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -256,88 +254,394 @@ def _format_multirow(
+    """Concrete table builder for regular table."""
+
+    @property
+    def env_begin(self):
```
try to type *everything*

ivanovmg replied:
Did that.
But faced mypy errors, that I do not know how to overcome.

``
pandas/io/formats/latex.py:44: error: Argument 1 to "list" has incompatible type "Iterator[Tuple[Any, ...]]"; expected "Iterable[List[str]]"  [arg-type]
pandas/io/formats/latex.py:544: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "str")  [assignment]
``
mypy does not know, for some reason, that list can get an iterator rather than iterable.

Another thing is related to ``column_format`` setter, that I do not know how to deal with.

Can you help?

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -256,88 +254,394 @@ def _format_multirow(
-        buf : string or file handle
-            File path or object. If not specified, the result is returned as
-            a string.
+    @property
```
this should be a class method no?

ivanovmg replied:
Not necessarily. At least not according to this https://refactoring.guru/design-patterns/builder/python/example

jreback commented on pandas/io/formats/latex.py:
```diff
@@ -256,88 +254,394 @@ def _format_multirow(
+                self._get_index_format() + self._get_column_format_based_on_dtypes()
+            )
+        elif not isinstance(input_column_format, str):  # pragma: no cover
+            raise AssertionError(
```
is this hit in the tests?

ivanovmg replied:
This is the original implementation.
I have just replaced AssertionError with ValueError and added several a test on this.

jreback reviewed (changes requested):
generally this is easier to follow. It feels slightly java-esque from the naming and builder pattern., but not too overboard. ok on the inheritance; almost always prefer composition, but this is not crazy. some comments.

ivanovmg committed:
Rename _ilevels, _nlevels, _clevels as reviewed

ivanovmg committed:
Move static methods to module level funcs

ivanovmg committed:
Add test for incorrect column_format

ivanovmg committed:
Add type annotations

Problems
--------

Could not figure out how to deal with the following.

1. Property setters
2. Inability of mypy to figure out that list can get iterator as
   argument rather than iterable only.

pandas/io/formats/latex.py:44: error: Argument 1 to "list" has incompatible type "Iterator[Tuple[Any, ...]]"; expected "Iterable[List[str]]"  [arg-type]
pandas/io/formats/latex.py:544: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "str")  [assignment]

ivanovmg committed:
Remove unnecessary list call

ivanovmg committed:
Drop comment

ivanovmg committed:
Set more descriptive _xlevels names

ivanovmg committed:
Ignore two problematic type checks

ivanovmg commented:
@jreback, is there anything else to be done from my side? It seems that I covered all the points you raised previously.

jreback reviewed (changes requested):
a couple of comments, pls merge master and ping on green.

ivanovmg committed:
REF: rename product property to get_result()

ivanovmg commented on pandas/io/formats/latex.py:
```diff
@@ -256,88 +216,449 @@ def _format_multirow(
+            position=self.position,
+        )
+
+    def _select_builder(self) -> Type[TableBuilderAbstract]:
```
I had to extract this method to make ``mypy`` happy.

If I refactor to this (which was originally there).

```
    @property
    def builder(self) -> TableBuilderAbstract:
         kwargs = dict(
            formatter=self.fmt,
            column_format=self.column_format,
            multicolumn=self.multicolumn,
            multicolumn_format=self.multicolumn_format,
            multirow=self.multirow,
            caption=self.caption,
            label=self.label,
            position=self.position,
        )
        if self.longtable:
            return LongTableBuilder(**kwargs)
        if any([self.caption, self.label, self.position]):
            return RegularTableBuilder(**kwargs)
        return TabularBuilder(**kwargs)
```
then ``mypy`` complains with the following error.

```
pandas/io/formats/latex.py:578: error: Argument 1 to "LongTableBuilder" has incompatible type "**Dict[str, object]"; expected "DataFrameFormatter"  [arg-type]
pandas/io/formats/latex.py:578: error: Argument 1 to "LongTableBuilder" has incompatible type "**Dict[str, object]"; expected "Optional[str]"  [arg-type]
pandas/io/formats/latex.py:578: error: Argument 1 to "LongTableBuilder" has incompatible type "**Dict[str, object]"; expected "bool"  [arg-type]
pandas/io/formats/latex.py:580: error: Argument 1 to "RegularTableBuilder" has incompatible type "**Dict[str, object]"; expected "DataFrameFormatter"  [arg-type]
pandas/io/formats/latex.py:580: error: Argument 1 to "RegularTableBuilder" has incompatible type "**Dict[str, object]"; expected "Optional[str]"  [arg-type]
pandas/io/formats/latex.py:580: error: Argument 1 to "RegularTableBuilder" has incompatible type "**Dict[str, object]"; expected "bool"  [arg-type]
pandas/io/formats/latex.py:581: error: Argument 1 to "TabularBuilder" has incompatible type "**Dict[str, object]"; expected "DataFrameFormatter"  [arg-type]
pandas/io/formats/latex.py:581: error: Argument 1 to "TabularBuilder" has incompatible type "**Dict[str, object]"; expected "Optional[str]"  [arg-type]
pandas/io/formats/latex.py:581: error: Argument 1 to "TabularBuilder" has incompatible type "**Dict[str, object]"; expected "bool"  [arg-type]
```

Honestly, typing stuff makes life more complicated, IMHO.

jreback replied:
yeap typing does make things more complicated, no arg there :->

ivanovmg committed:
REF: rename base builder, add doctests

 - TableBuilder -> GenericTableBuilder
 - Add doctests explaining what each table builder does

ivanovmg committed:
Merge branch 'master' into refactor/to_latex

ivanovmg commented on pandas/io/formats/latex.py:
```diff
@@ -256,88 +216,449 @@ def _format_multirow(
+        self.position = position
+
+    @property
+    def product(self) -> str:
```
Changed to a method ``get_result()``.

ivanovmg commented on pandas/io/formats/latex.py:
```diff
@@ -256,88 +216,449 @@ def _format_multirow(
+        return "\n".join(["\\end{tabular}", "\\end{table}"])
+
+
+class TabularBuilder(TableBuilder):
```
Renamed ``TableBuilder -> GenericTableBuilder``. This is a parent class for all table builders. These include longtable, table and tabular. The last one is the non-floating environment.
The reason they are all separate is that they have slightly different behavior.

```
\begin{tabular}
...
\end{tabular}
```

I added doctests explaining what each of the builders tries to accomplish.
Would that be sufficient enough?

ivanovmg committed:
TYP: ignore specific mypy errors

pep8speaks commented:
Hello @ivanovmg! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/io/formats/latex.py`](https://github.com/pandas-dev/pandas/blob/94d8dc72ea2e3dd4d8d0ef6fbeac4e1e3900ac77/pandas/io/formats/latex.py):

> [Line 44:89](https://github.com/pandas-dev/pandas/blob/94d8dc72ea2e3dd4d8d0ef6fbeac4e1e3900ac77/pandas/io/formats/latex.py#L44): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 88 characters)

ivanovmg committed:
FIX: long line with type ignore

jreback reviewed (approved):
lgtm ping on green.

ivanovmg commented:
Can anybody restart the task Windows py37_np16?
It is unlikely that the failure in ``TestPandasContainer.test_to_s3`` is anyhow related to the present PR.

jreback commented:
thanks @ivanovmg for this, very nice!
